content/en/docs/components/pipelines/author-a-pipeline/inputs-outputs-data-passing.md

```diff
@@ -0,0 +1,349 @@
+
+```python
+from kfp import dsl
+from kfp.dsl import Input, Output, Dataset
+
+
+@dsl.container_component
+def create_dataset(
+    initial_text: str,
+    output_dataset: Output[Dataset],
+):
+    """Create a dataset containing the string `initial_text`.""" 
+    return dsl.ContainerSpec(
+        image='alpine',
+        command=['sh', '-c', 'mkdir --parents $(dirname "$1") && echo "$0" > "$1"',],
+        args=[initial_text, output_dataset.path])
+
+
+@dsl.component
+def augment_dataset(
+    existing_dataset: Input[Dataset],
+    resulting_dataset: Output[Dataset],
+    text: str,
+    num: int = 10,
+) -> int:
+    """Append `text` `num` times to an existing dataset, then write it as a new dataset."""
+    additional_data = ' '.join(text for _ in range(num))
+
+    with open(existing_dataset.path, 'r') as f:
+        existing_dataset_text = f.read()
+
+    resulting_dataset_text = existing_dataset_text + ' ' + additional_data
+
+    with open(resulting_dataset.path, 'w') as f:
+        f.write(resulting_dataset_text)
+
+    return len(resulting_dataset_text)
+
+
+@dsl.pipeline()
+def my_pipeline(initial_text: str = 'initial dataset text'):
+    create_task = create_dataset(initial_text=initial_text)
+    augment_dataset(
+        existing_dataset=create_task.outputs['output_dataset'],
+        text='additional text')
+```
+
+This pipeline uses a container component `create_dataset` to construct an initial `Dataset` artifact containing `initial_text`. Then, the downstream lightweight Python component `augment_dataset` appends `text` repeated `num` times to the dataset and saves it as a new dataset.
+
+## Inputs
+Component inputs are specified by the component function's signature. This applies for all authoring approaches: [lightweight Python components][lightweight-python-component], [containerized Python components][containerized-python-component], and [container components][container-component].
+
+Ultimately, each authoring style creates a component definitied by an `image`, `command`, and `args`. When you use an input, it is represented as a placeholder in the `command` or `args` and is interpolated at component runtime.
+
+There is one additional type of input, `PipelineTaskFinalStatus`, which allows access to the metadata of one task from within another. This input is a special case, as it is neither a parameter nor an artifact. Use of this input is covered in [Authoring: Pipelines][pipelines].
+
+### Input parameters
+Input parameters are declared when you use a `str`, `int`, `float`, `bool`, `dict` or `list` type annotation. The data passed to parameters typed with `dict` or `list` may only container JSON-serializable Python primitives. Inner types to `dict` and `list`, such as `Dict[str, str]` or `List[int]` are not respected. `Union` types are not permitted.
```

Comment:
> Inner types to `dict` and `list`, such as `Dict[str, str]` or `List[int]` are not respected.

IIRC, in some cases they are even required.
https://github.com/kubeflow/pipelines/blob/efb96135033fc6e6e55078d33814c45a98566e68/sdk/python/kfp/compiler/test_data/pipelines/pipeline_with_loops.py#L23

Reply:
You are correct. Updated.